hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/systemcontracts/PrngSystemContract.java

```diff
@@ -0,0 +1,138 @@
+ */
+@Singleton
+public class PrngSystemContract extends AbstractPrecompiledContract {
+    private static final String PRECOMPILE_NAME = "PRNG";
+    // random256BitGenerator(uint256)
+    static final int PSEUDORANDOM_SEED_GENERATOR_SELECTOR = 0xd83bf9a1;
+    public static final String PRNG_PRECOMPILE_ADDRESS = "0x169";
+    private long gasRequirement;
+
+    @Inject
+    public PrngSystemContract(@NonNull final GasCalculator gasCalculator) {
+        super(PRECOMPILE_NAME, gasCalculator);
+    }
+
+    @Override
+    public long gasRequirement(final Bytes bytes) {
+        return gasRequirement;
+    }
+
+    @Override
+    public @NonNull PrecompileContractResult computePrecompile(final Bytes input, @NonNull final MessageFrame frame) {
+        requireNonNull(input);
+        requireNonNull(frame);
+
+        // compute the gas requirement
+        gasRequirement =
+                calculateGas(Instant.ofEpochSecond(frame.getBlockValues().getTimestamp()));
+
+        // get the contract ID
+        final ContractID contractID = asEvmContractId(Address.fromHexString(PRNG_PRECOMPILE_ADDRESS));
+
+        try {
+            // compute the pseudorandom number
+            final var randomNum = generatePseudoRandomData(input, frame);
+            requireNonNull(randomNum);
+            final var result = PrecompiledContract.PrecompileContractResult.success(randomNum);
+
+            // create a child record if we are not in a static call
+            if (!frame.isStatic()) {
+                createSuccessfulRecord(frame, randomNum, contractID);
+            }
+
+            return result;
+        } catch (Exception e) {
+            // create a failed record and returned a halt result
+            createFailedRecord(frame, FAIL_INVALID.toString(), contractID);
+            return PrecompiledContract.PrecompileContractResult.halt(Bytes.EMPTY, Optional.of(INVALID_OPERATION));
+        }
+    }
+
+    void createSuccessfulRecord(
+            @NonNull MessageFrame frame, @NonNull final Bytes randomNum, @NonNull final ContractID contractID) {
+        requireNonNull(frame);
+        requireNonNull(randomNum);
+        requireNonNull(contractID);
+        var updater = (ProxyWorldUpdater) frame.getWorldUpdater();
+        updater.externalizeSystemContractResults(
+                contractFunctionResultSuccessFor(gasRequirement, randomNum, contractID), false);
+    }
+
+    void createFailedRecord(
+            @NonNull MessageFrame frame, @NonNull final String errorMsg, @NonNull final ContractID contractID) {
+        requireNonNull(frame);
+        requireNonNull(contractID);
+        contractFunctionResultFailedFor(gasRequirement, errorMsg, contractID);
+        var updater = (ProxyWorldUpdater) frame.getWorldUpdater();
+        updater.externalizeSystemContractResults(
+                contractFunctionResultFailedFor(gasRequirement, errorMsg, contractID), true);
```

Comment:
It is kind of annoying (and misleading!) that `ProxyWorldUpdater.externalizeSystemContractResults` (and `SystemContractOperations.externalizeResult` take a _boolean_ `isError` which is _backwards_: you have to pass `false` for the success case (above at L102) and `true` for the failure case (here).  A good example of why passing a bare boolean is frequently not the best approach.  I suggest creating an enum (over there at `SystemContractOperations.externalizeResult` and using it as the argument type.  Perhaps `enum Result { IS_SUCCESS, IS_IN_ERROR }`

Reply:
BTW I just discovered this branch doesn't build for me, I get errors in `mono/state/migration/{AccountStateTranslator,TokenStateTranslator}.java`.  You're not seeing that?

Reply:
Must have been something in my setup then.